### PR TITLE
Packages: Extract @wordpress/scripts-utils package

### PR DESCRIPTION
**Pull request by gziolo**
## Description

New npm package `@wordpress/scripts-utils` that exposed general-purpose utilities to use with other packages. Methods included:

- `fromProjectRoot`
- `getArgFromCLI`
- `getArgsFromCLI`
- `getFileArgsFromCLI`
- `hasArgInCLI`
- `hasFileArgInCLI`
- `hasPackageProp`
- `hasProjectFile`

**I plan to document all those methods separately together with TypeScript in JSDoc integration.** 

## How has this been tested?

Travis should be green and all existing `wp-scripts` should work as before. 

In fact, we use almost all features included in `wp-scripts` so it self-tests :)
## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate.

### CODE REVIEW

File: packages/eslint-plugin/configs/recommended.js [hunk 1]
```diff
@@ -1,7 +1,27 @@
...
+	config.rules = {
+		'prettier/prettier': [
+			'error',
+			defaultPrettierConfig,
+			{
+				usePrettierrc: false,
+			},
+		],
+	};
+}
```
Comment (aduth) [inline packages/eslint-plugin/configs/recommended.js:25]: I see what it is that we're trying to address, but something doesn't sit very well with me that the configuration is generated dynamically based on factors from the environment (directory) from which the linting is run. Is there any prior art to this?

**Comment by aduth** [comment]
I'm typically not opposed to breaking down shared code to small shared libraries. I do feel the use-cases are pretty limited and solving very specific problems. There's probably not much usability outside these modules? I could see it being more useful if the methods (and the package itself) were targeted toward more generic "does X file exist in the project directory", "is X given as a CLI argument". Even these are pretty unique / distinct. Some of them exist here, but then there are others which are _very_ specific like `hasPrettierConfig`.

Also, I've been trying unsuccessfully to kill off `camelCaseDash` for years, and it only seems to propagate more over time 😄

**Comment by gziolo** [comment]
> There's probably not much usability outside these modules?

Well, I can’t counter argument this 😃

> Some of them exist here, but then there are others which are _very_ specific like `hasPrettierConfig`.

The one mentioned is the most important one.

> Also, I've been trying unsuccessfully to kill off `camelCaseDash` for years, and it only seems to propagate more over time 😄 (related: https://github.com/WordPress/gutenberg/pull/14869#discussion_r273472691, https://github.com/WordPress/gutenberg/pull/13814/files#r258035009)

As noted, it plays its role in a few places for better or worse 😅

File: packages/eslint-plugin/configs/recommended.js (hunk 1 above)
Comment (gziolo) [reply packages/eslint-plugin/configs/recommended.js:25]: How about we simply make this ESLint rule always present with this custom WordPress Prettier config set? If someone doesn’t like it, they can use a different ruleset offered or simply override it.

In fact, I could do the opposite in the `wp-scripts lint-js` to what this PR tries and override the rule only when the Prettier config is detected in the project.

**Comment by kansas1** [comment]
👍

File: packages/eslint-plugin/configs/recommended.js (hunk 1 above)
Comment (aduth) [reply packages/eslint-plugin/configs/recommended.js:25]: > How about we simply make this ESLint rule always present with this custom WordPress Prettier config set? If someone doesn’t like it, they can use a different ruleset offered or simply override it.

It seems reasonable to me. As I consider it more, I think it's pretty sensible that these rules assume a default that the Prettier rules be configured to the WordPress standard. I'm pretty okay if we keep consideration of an override configuration from the local directory, if it helps avoid a breaking change.

Comment (aduth) [reply packages/eslint-plugin/configs/recommended.js:25]: >How about we simply make this ESLint rule always present with this custom WordPress Prettier config set?

Would this mean we wouldn't need `hasPrettierConfig` here anymore? Which I think is largely the primary purpose of the proposed new package, in making this utility available? Per https://github.com/WordPress/gutenberg/pull/21503#issuecomment-612893121, I think I wouldn't really mind some of these utilities still. I'd be happier if we didn't have `hasBabelConfig`, `hasJestConfig`, `hasPrettierConfig`, and `camelCaseDash` on the public API though.

Comment (gziolo) [reply packages/eslint-plugin/configs/recommended.js:25]: > I think it's pretty sensible that these rules assume a default that the Prettier rules be configured to the WordPress standard.

> Per [#21503 (comment)](https://github.com/WordPress/gutenberg/pull/21503#issuecomment-612893121), I think I wouldn't really mind some of these utilities still. I'd be happier if we didn't have `hasBabelConfig`, `hasJestConfig`, `hasPrettierConfig`, and `camelCaseDash` on the public API though.

It all makes sense. Let's do it and open another PR that addresses the related issue with the default coding styles set for Prettier in ESLint.

**Comment by gziolo** [comment]
@aduth, I opened https://github.com/WordPress/gutenberg/pull/21602 to address the part for Prettier. It turned out to be a very elegant solution ✨

File: packages/eslint-plugin/configs/recommended.js (hunk 1 above)
Comment (gziolo) [reply packages/eslint-plugin/configs/recommended.js]: Prettier + ESLint integration was tackled in https://github.com/WordPress/gutenberg/pull/21602.

I updated this PR according to the feedback received. It's ready for another check.

**Comment by aduth** [comment]
Something still isn't sitting quite well with me about this. I think it's my dislike of the term "utils", which to me almost always may as well be substituted with "stuff". I think for me, I'd prefer if we'd do one of:

- Either figure out what this "stuff" is meant to represent, and name it as such.
- Or keep it internal to `@wordpress/scripts`, if it really is just _script's stuff_. 

As an extension of previous conversations, I think I can see how it could be organized in a way where we could have one (or even more) packages for:

- CLI argument interpretation
- Extracting values from `package.json`
- Checking for presence of files in a particular directory
   - And/or resolving from a "root" folder

**Comment by gziolo** [comment]
>  I think it's my dislike of the term "utils", which to me almost always may as well be substituted with "stuff" as assorted miscellany. And I don't think assorted miscellany makes for a good package.

I have to agree with you. In the past, I disbanded `wp.utils` (before packages) and I split it into small packages mostly because it was a random group of utils that didn’t fit anywhere else... I have sometimes a similar thoughts about `@wordpress/components` that became a package to go when contributors want to share code. I’m seeing now another package that starts to play a similar role: `@wordpress/interface`

I like your proposal of how those utilities could be divided. I guess it’s find to wait until we really want to reuse this functionality in Gutenberg.

Thank you for all the feedback, let’s close this PR. I merged 2 PRs regardless with your help that solve some of the issues that triggered this exploration 😃

**Pull request closed without merging**